### ISSUE TITLE
Exception when using a custom model in the Fiji plugin StarDist

### ISSUE THREAD

**Issue by Chiara-Sgat** [HenriquesLab/ZeroCostDL4Mic#216]
Hi,

I have trained a StarDist model in GoogleColab using the stardist_zerocostDL4mic notebook. Once I tried to use the .zip file in the Fiji plugin StarDist on another image, it shows up this exception:

(Fiji Is Just) ImageJ 2.3.0/1.53f51; Java 1.8.0_172 [64-bit]; Windows 10 10.0; 64MB of 5935MB (1%)

java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception at net.imagej.legacy.LegacyService.runLegacyCompatibleCommand(LegacyService.java:308)

at net.imagej.legacy.DefaultLegacyHooks.interceptRunPlugIn(DefaultLegacyHooks.java:166)
at ij.IJ.runPlugIn(IJ.java)
at ij.Executer.runCommand(Executer.java:150)
at ij.Executer.run(Executer.java:68)
at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
at java.util.concurrent.FutureTask.report(FutureTask.java:122)
at java.util.concurrent.FutureTask.get(FutureTask.java:192)
at net.imagej.legacy.LegacyService.runLegacyCompatibleCommand(LegacyService.java:304)
… 5 more
Caused by: java.lang.RuntimeException: Module threw exception
at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
… 1 more
Caused by: java.lang.NullPointerException
at de.csbdresden.stardist.StarDist2D.splitPrediction(StarDist2D.java:338)
at de.csbdresden.stardist.StarDist2D.run(StarDist2D.java:307)
at org.scijava.command.CommandModule.run(CommandModule.java:196)
at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
… 6 more

I have tried to downgrade TensorFlow to the version 1.15.0 and 1.14.0 before the training and change the version in Fiji too but nothing has changed. 

Thanks in advance for you help,
Chiara

**Comment by esgomezm** [comment]
Dear Chiara, 

Thank you for trying it. We updated the notebook to TensorFlow version 2.8 so it could connect with google colab's GPU. However, this compromises the compatibility with Fiji, as it only reads models until TensorFlow 2.2. Fiji DL libraries (CSDeep and deepImageJ) are expected to be updated soon but mean while you can try the following:

At the beginning of the notebook, in the code cell after 1.1. Install key dependencies, you have the following code:

````python
#@markdown ##Install StarDist and dependencies
# Install the correct TF version for Google Colab
#!pip uninstall -y -q tensorflow
#!pip install -q tensorflow-probability==0.8
#!pip install -q kapre==0.1.7
#!pip install -q tensorflow==2.2


# Install packages which are not included in Google Colab
!pip install -q tifffile # contains tools to operate tiff-files
!pip install -q csbdeep  # contains tools for restoration of fluorescence microcopy images (Content-aware Image Restoration, CARE). It uses Keras and Tensorflow.
!pip install 'stardist[bioimageio]'
#!pip install -q stardist # contains tools to operate STARDIST.
!pip install -q gputools # improves STARDIST performances
#!pip install edt # improves STARDIST performances
!pip install -q wget
!pip install -q fpdf
!pip install -q PTable # Nice tables 
!pip install -q zarr
!pip install -q imagecodecs
#!pip uninstall -y -q keras-nightly


#Force session restart
#exit(0)
````
Try uncommenting the first lines as follows and run the cell. Wait until it finishes and the dialog in the bottom part appears saying the the session has been restarted.

````python
#@markdown ##Install StarDist and dependencies
# Install the correct TF version for Google Colab
!pip uninstall -y -q tensorflow
!pip install -q tensorflow-probability==0.8
!pip install -q kapre==0.1.7
!pip install -q tensorflow==2.2


# Install packages which are not included in Google Colab
!pip install -q tifffile # contains tools to operate tiff-files
!pip install -q csbdeep  # contains tools for restoration of fluorescence microcopy images (Content-aware Image Restoration, CARE). It uses Keras and Tensorflow.
!pip install 'stardist[bioimageio]'
#!pip install -q stardist # contains tools to operate STARDIST.
!pip install -q gputools # improves STARDIST performances
#!pip install edt # improves STARDIST performances
!pip install -q wget
!pip install -q fpdf
!pip install -q PTable # Nice tables 
!pip install -q zarr
!pip install -q imagecodecs
#!pip uninstall -y -q keras-nightly


#Force session restart
exit(0)
````

Then, go to section 5 directly and try loading the model you already had trained. If it's able to load, export it to the bioimage model zoo format and you should be able to use it in Fiji. 
If it gives you errors when loading the model, then the only solution I see at the moment would be to run the training again with this new configuration. Take into account that it will be much slower than the previous one as you won't have GPU connection. 

I hope this helps

**Comment by Chiara-Sgat** [comment]
Hi,

Thank you for your reply and help! 

I have been trying it and I finally managed to get a result.
I have tried to uncomment the first lines in the code cell 1.1 as you suggested but I got this error:

_ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
csbdeep 0.7.2 requires h5py<3; python_version < "3.9", but you have h5py 3.7.0 which is incompatible.
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow-probability 0.8.0 requires gast<0.3,>=0.2, but you have gast 0.3.3 which is incompatible._

However, I have tried to install the following versions and I didn't get any error anymore:
-gast=0.3.3.
-h5py=2.10.0
-cloudpickle=1.5.0
-TensorFlow-probability=0.9
-TensorFlow=2.2.0

I went on with the training of the data (using the CPU) and everything worked great. 
Then I exported it to the bioimage model zoo format and installed it and ran it using DeepImageJ, and it worked great after changing the TensorFlow version of Fiji to 1.14 CPU.

I am not really into models but it seems that these changes did not affect the general performance of the training. 

Thank you again for your help,
Chiara
